subprojects/core/src/main/java/org/gradle/api/internal/project/AllprojectsAwareProject.java

```diff
@@ -0,0 +1,1172 @@
+import org.gradle.api.internal.plugins.ExtensionContainerInternal;
+import org.gradle.api.internal.plugins.PluginManagerInternal;
+import org.gradle.api.internal.tasks.TaskContainerInternal;
+import org.gradle.api.internal.tasks.TaskDependencyFactory;
+import org.gradle.api.logging.Logger;
+import org.gradle.api.logging.LoggingManager;
+import org.gradle.api.model.ObjectFactory;
+import org.gradle.api.plugins.ObjectConfigurationAction;
+import org.gradle.api.plugins.PluginContainer;
+import org.gradle.api.project.IsolatedProject;
+import org.gradle.api.provider.Property;
+import org.gradle.api.provider.Provider;
+import org.gradle.api.provider.ProviderFactory;
+import org.gradle.api.resources.ResourceHandler;
+import org.gradle.api.tasks.WorkResult;
+import org.gradle.configuration.ConfigurationTargetIdentifier;
+import org.gradle.configuration.project.ProjectConfigurationActionContainer;
+import org.gradle.groovy.scripts.ScriptSource;
+import org.gradle.internal.logging.StandardOutputCapture;
+import org.gradle.internal.metaobject.DynamicObject;
+import org.gradle.internal.model.ModelContainer;
+import org.gradle.internal.model.RuleBasedPluginListener;
+import org.gradle.internal.service.ServiceRegistry;
+import org.gradle.internal.service.scopes.ServiceRegistryFactory;
+import org.gradle.invocation.IsolatedProjectEvaluationListenerProvider;
+import org.gradle.model.internal.registry.ModelRegistry;
+import org.gradle.normalization.InputNormalizationHandler;
+import org.gradle.normalization.internal.InputNormalizationHandlerInternal;
+import org.gradle.process.ExecResult;
+import org.gradle.process.ExecSpec;
+import org.gradle.process.JavaExecSpec;
+import org.gradle.util.Path;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.io.File;
+import java.net.URI;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.Callable;
+
+public class AllprojectsAwareProject extends GroovyObjectSupport implements ProjectInternal, DynamicObjectAware {
+
+    private final ProjectInternal delegate;
+    private final IsolatedProjectEvaluationListenerProvider isolatedProjectEvaluationListenerProvider;
+    private final GradleInternal gradle;
+
+    private boolean isAllprojectsActionInitialized = false;
+    private IsolatedAction<? super ProjectInternal> allprojectsAction = null;
+
+    public AllprojectsAwareProject(
+        ProjectInternal delegate,
+        IsolatedProjectEvaluationListenerProvider isolatedProjectEvaluationListenerProvider,
+        GradleInternal gradle
+    ) {
+        this.delegate = delegate;
+        this.isolatedProjectEvaluationListenerProvider = isolatedProjectEvaluationListenerProvider;
+        this.gradle = gradle;
+    }
+
+    private void applyAllprojects() {
+        if (!isAllprojectsActionInitialized) {
+            allprojectsAction = isolatedProjectEvaluationListenerProvider.isolateAllprojectsActionFor(gradle);
```

Comment:
❓ Why do we store the action in the field? If I read it right, it might be executed multiple times for the same project.

Reply:
No, it shouldn't be called multiple times. Do you see a reason for this?
We store it to avoid extra allocation from call `isolatedProjectEvaluationListenerProvider.isolateAllprojectsActionFor`

Reply:
If `applyAllProjects()` is called two times, it will execute the action two times, no?

> We store it to avoid extra allocation from call
Why do you need to store it at all, if you only need to run it once? Why not store just a boolean?

Reply:
> If applyAllProjects() is called two times, it will execute the action two times, no?
No, [see](https://github.com/gradle/gradle/pull/29611/files#diff-19893253a85e7237faf3ad9fc7b7385a2a1b8240294e2c32f90d9abb22ce8da5R148)

> Why not store just a boolean?
Ah, you mean this, we can do it, yes

Reply:
Then it will simplify the state within the class

Reply:
> No, [see](https://github.com/gradle/gradle/pull/29611/files#diff-19893253a85e7237faf3ad9fc7b7385a2a1b8240294e2c32f90d9abb22ce8da5R148)

Oh, I see. The action does run twice or more, we just abort it from within.

But it would indeed be nicer to avoid dispatching it more than once from the same project wrapper.